### PR TITLE
feat: add penalty reference details page

### PR DESCRIPTION
**Pull request by akaJSaunders**
### Jira Link
https://companieshouse.atlassian.net/browse/LFA-1262

### Description
Penalty details entry page with page and input validation implemented. Integrations tests for http requests and unit testing for validation present.

### Work checklist

- [x] Tests added where applicable
- [ ] UI changes look good on mobile
- [x] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achive this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by geek-soft** [comment]
What is the purpose of having `dump.rdb` file in the repository?

File: .env.development [hunk 1]
```diff
@@ -0,0 +1,2 @@
+REDIS_HOST=<HOST_HERE>
```
Comment (geek-soft) [inline .env.development]: Instead of `<HOST_HERE>` and so one real values should be used. If this is for vagrant environment then I suggest you point it into host and port provided by the environment.

File: package.json [hunk 2]
```diff
@@ -18,11 +18,16 @@
...
+    "@hapi/hapi": "^19.1.0",
+    "@hapi/joi": "^17.1.0",
+    "@types/joi": "^14.3.4",
     "body-parser": "^1.18.2",
     "cookie-parser": "^1.4.4",
     "dotenv": "^8.2.0",
     "express": "^4.15.5",
+    "express-joi-validation": "^4.0.3",
     "express-session": "^1.15.6",
+    "express-validator": "^6.4.0",
```
Comment (geek-soft) [inline package.json]: I thought you are not going to use it anymore? Shouldn't that be removed?

File: package.json [hunk 3]
```diff
@@ -18,11 +18,16 @@
...
   },
   "dependencies": {
+    "@hapi/hapi": "^19.1.0",
+    "@hapi/joi": "^17.1.0",
+    "@types/joi": "^14.3.4",
     "body-parser": "^1.18.2",
     "cookie-parser": "^1.4.4",
     "dotenv": "^8.2.0",
     "express": "^4.15.5",
+    "express-joi-validation": "^4.0.3",
```
Comment (geek-soft) [inline package.json]: Seems like something never user? I suggest you remove it.

File: package.json [hunk 4]
```diff
@@ -40,6 +45,7 @@
     "@types/dotenv": "^8.2.0",
     "@types/express": "^4.0.37",
     "@types/express-session": "^1.15.3",
+    "@types/hapi__joi": "^16.0.9",
```
Comment (geek-soft) [inline package.json]: What is the difference between `@types/hapi__joi` and `@types/joi`? Isn't there an overlap?

File: package.json [hunk 5]
```diff
@@ -18,11 +18,16 @@
     "test": "mocha test/**/*.test.ts"
   },
   "dependencies": {
+    "@hapi/hapi": "^19.1.0",
+    "@hapi/joi": "^17.1.0",
+    "@types/joi": "^14.3.4",
```
Comment (geek-soft) [inline package.json]: All types should be defined as dev dependencies.

File: src/models/PenaltyReferenceDetails.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+export class PenaltyReferenceDetails  {
```
Comment (geek-soft) [inline src/models/PenaltyReferenceDetails.ts]: Technically that could be just an interface cause currently class does not add any behaviour, it is just data bucket.

File: src/services/SessionService.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import { provide } from 'inversify-binding-decorators';
+import { Session } from '../models/Session';
+import * as crypto from 'crypto';
+import { IMap } from 'src/models/types';
+import { RedisService } from './RedisService';
+import { PenaltyReferenceDetails } from '../models/PenaltyReferenceDetails';
+
+
+@provide(SessionService)
+export class SessionService {
```
Comment (geek-soft) [inline src/services/SessionService.ts]: This will be thrown away so I suggest you do not spend much more time polishing it.

File: src/utils/Schemas.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        .uppercase()
+        .regex(/([A-Z]{1}[0-9]{8})/)
+        .messages({
+            'string.empty': 'You must enter a penalty reference number',
+            'string.pattern.base': 'You must enter your reference number exactly as shown on your penalty notice'
+        })
+});
+
+
+export const padNumber = (companyNumber: string): string => {
```
Comment (geek-soft) [inline src/utils/Schemas.ts]: In my opinion that method does not belong to `Schemas.ts`. It is more utility class for company numbers than anything related to schemas.

File: src/views/penaltydetails.njk [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+  {{ 
+    govukPhaseBanner({
+      tag: {
+        text: "alpha"
+      },
+      html: 'This is a new service – your <a class="govuk-link" href="#">feedback</a> will help us to improve it.'
+    }) 
+  }}
+  {{ 
+    govukBackLink({
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: Shall we quickly remove it as it is about of scope of the story? Same as phase banner?

File: package.json (hunk 2 above)
Comment (n-r-b) [reply package.json]: Removed this

File: package.json (hunk 3 above)
Comment (n-r-b) [reply package.json]: Removed this

File: package.json (hunk 4 above)
Comment (n-r-b) [reply package.json]: Looks like we are not using `@types/hapi__joi`, so i have removed this

File: src/views/penaltydetails.njk [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+              text: "This is the 9-character reference shown on the penalty notice."
+            },
+            errorMessage: validationResult.getErrorForField('penaltyReference') if validationResult,
+            classes: "govuk-input--penalty-reference",
+            id: "penalty-reference",
+            name: "penaltyReference",
+            value: penaltyReference
+          }) 
+        }}
+        <div class="govuk-form-group">
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: This is out of scope of the story.

File: src/views/penaltydetails.njk [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+              summaryText: "Help with company number and penalty reference number",
+              html: detailsHTML
+            }) 
+          }}
+        </div>
+        {{ 
+          govukButton({
+            text: "Continue",
+            attributes: {
+              "data-event-id": "continue-button",
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: What is that `data-event-id` for?

File: src/views/penaltydetails.njk (hunk 9 above)
Comment (akaJSaunders) [reply src/views/penaltydetails.njk]: agreed, although the layout.njk includes a Beta banner. Do you suggest we remove it from there aswell?

File: src/utils/Schemas.ts (hunk 8 above)
Comment (akaJSaunders) [reply src/utils/Schemas.ts]: I am not very fond of the padNumber function, will look into how to simplify it

File: src/views/penaltydetails.njk (hunk 11 above)
Comment (n-r-b) [reply src/views/penaltydetails.njk]: I added this as i thought it was required for the id on the continue button. I'll remove this for now

File: src/utils/Paths.ts [hunk 12]
```diff
@@ -1,8 +1,14 @@
 export const ROOT = '';
 
+export const OTHER_REASON_PREFIX=`${ROOT}/other`
+export const OTHER_REASON_ENTRY_PREFIX=`/other-reason-entry`
+
+export const PENALTY_DETAILS_PREFIX=`${ROOT}/penalty-reference`
+export const PENALTY_DETAILS_PAGE_URL=`${PENALTY_DETAILS_PREFIX}`
```
Comment (geek-soft) [inline src/utils/Paths.ts]: It is never used, please remove.

File: src/views/penaltydetails.njk [hunk 13]
```diff
@@ -0,0 +1,68 @@
+{% extends "layout.njk" %}
+
+{% from "govuk/components/error-summary/macro.njk" import govukErrorSummary %}
+{% from "govuk/components/back-link/macro.njk"  import govukBackLink %}
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: Since you removed back link it make sense to remove macro as well.

File: src/views/penaltydetails.njk [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+{% from "govuk/components/input/macro.njk"      import govukInput %}
+{% from "govuk/components/details/macro.njk"    import govukDetails %}
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: Since you removed details block it make sense to remove macro import as well.

File: src/views/penaltydetails.njk [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+{% block pageTitle %}
+  Tell us the penalty details
+{% endblock %}
+
+{% block content %}
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+        {{ govukErrorSummary
+          ({
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: I suggest the following formatting for consistency:

image

File: src/views/penaltydetails.njk [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+            errorMessage: validationResult.getErrorForField('penaltyReference') if validationResult,
+            classes: "govuk-input--penalty-reference",
+            id: "penalty-reference",
+            name: "penaltyReference",
+            value: penaltyReference
+          }) 
+        }}
+        {{ 
+          govukButton({
+            text: "Continue",
```
Comment (geek-soft) [inline src/views/penaltydetails.njk]: We are not consistent in terms of using single vs double quotes. At some point we should settle on one of these. Not necessarily problem in the context of this story.

File: test/utils/validation/SchemaValidator.test.ts [hunk 17]
```diff
@@ -13,7 +14,7 @@ describe('SchemaValidator', () => {
         });
     });
 
-    describe('validation', () => {
+    describe('demo validation', () => {
```
Comment (geek-soft) [inline test/utils/validation/SchemaValidator.test.ts]: Please revert all the changes in the file. They are not needed.

File: test/controllers/PenaltyDetailsController.test.ts [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+const app = createApplication(container => {
+    container
+        .bind(RedisService).toConstantValue(createSubstituteOf<RedisService>(service => {
+            service.ping().returns(true)
+        }));
+});
+
+describe('PenaltyDetailsController', () => {
+
+    it('should return 200 when trying to access the penalty-reference page', async () => {
```
Comment (geek-soft) [inline test/controllers/PenaltyDetailsController.test.ts]: May I suggest you follow example set by 'test/controllers/OtherReasonController.test.ts', specifically:
- grouping GET and POST requests
- content assertions such as:
  - checking page heading
  - checking error summary heading
  - checking error messages

File: test/controllers/PenaltyDetailsController.test.ts [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+        await request(app)
+            .get('/penalty-reference')
+            .expect(OK);
+    });
+
+    // it('should return 200 when posting valid penalty details', async () => {
```
Comment (geek-soft) [inline test/controllers/PenaltyDetailsController.test.ts]: No code that is commented out please. Anything that does not run should be removed.

File: src/views/penaltydetails.njk (hunk 16 above)
Comment (akaJSaunders) [reply src/views/penaltydetails.njk]: my editor enforces single quotes in typescript files, not in njks (which are mostly copied from the prototype). But I think we are generally in agreement regarding single quotes

File: test/controllers/PenaltyDetailsController.test.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+    it('should return 400 when posting empty penalty reference', async () => {
+        const penaltyDetails: PenaltyReferenceDetails = {
+            penaltyReference: '',
+            companyNumber: 'SC123123'
+        };
+        await request(app)
+            .post('/penalty-reference')
+            .send(penaltyDetails)
+            .expect(BAD_REQUEST);
```
Comment (geek-soft) [inline test/controllers/PenaltyDetailsController.test.ts]: Please change it to 422 not 400. A 400 is for `The server could not understand the request due to invalid syntax.` while 422 is for `The request was well-formed but was unable to be followed due to logic errors.`. Request with invalid form data are valid from the syntax standpoint, they are just not acceptable from an application point of view.

Other definitions:

400 - The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.

422 - The 422 (Unprocessable Entity) status code means the server understands the content type of the request entity (hence a 415(Unsupported Media Type) status code is inappropriate), and the syntax of the request entity is correct (thus a 400 (Bad Request) status code is inappropriate) but was unable to process the contained instructions. For example, this error condition may occur if an XML request body contains well-formed (i.e., syntactically correct), but semantically erroneous, XML instructions.

Comment (n-r-b) [reply test/controllers/PenaltyDetailsController.test.ts]: 👍 - thanks for the explanation,  i've updated this

File: src/models/PenaltyReferenceDetails.schema.ts [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+    companyNumber: Joi.string()
+        .required()
+        .regex(/^(((SC|NI)[0-9]{6})|([0-9]{8}))$/)
+        .messages({
+            'string.empty': 'You must enter a company number',
+            'string.pattern.base': 'You must enter your full eight character company number'
+        }),
+    penaltyReference: Joi.string()
+        .required()
+        .regex(/([A-Z]{1}[0-9]{8})$/)
```
Comment (geek-soft) [inline src/models/PenaltyReferenceDetails.schema.ts]: Story says that case is not relevant so it could be A-Z or a-z.

File: test/models/PenaltyEntrySchema.test.ts [hunk 22]
```diff
@@ -10,240 +9,173 @@ describe('Penalty Details Schema Validation', () => {
...
-                const penaltyDetails: PenaltyReferenceDetails = {
-                    penaltyReference: 'b12345678',
-                    companyNumber: 'NI123123'
-                };
-                const result = validator.validate(penaltyDetails)
-                const expectedResult = {
-                    errors:[{field: 'penaltyReference', text: 'You must enter your reference number exactly as shown on your penalty notice'}]
-                }
-                expect(result).to.deep.equal(expectedResult)
+                const result = validator.validate(createModelWithPenaltyReference('b12345678'));
```
Comment (n-r-b) [inline test/models/PenaltyEntrySchema.test.ts]: this case should be valid according to the  ticket @akaJSaunders

Comment (akaJSaunders) [reply test/models/PenaltyEntrySchema.test.ts]: Well spotted

Comment (geek-soft) [reply test/models/PenaltyEntrySchema.test.ts]: Changed in most recent version.
